zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementMigratedApplier.java

```diff
@@ -43,9 +43,9 @@ public void applyState(final long elementInstanceKey, final ProcessInstanceRecor
 
     if (value.getBpmnElementType() == BpmnElementType.PROCESS) {
       elementInstanceState.deleteProcessInstanceKeyByDefinitionKey(
-          value.getProcessInstanceKey(), previousProcessDefinitionKey.get());
+          value.getTenantId(), value.getProcessInstanceKey(), previousProcessDefinitionKey.get());
       elementInstanceState.insertProcessInstanceKeyByDefinitionKey(
-          value.getProcessInstanceKey(), value.getProcessDefinitionKey());
+          value.getTenantId(), value.getProcessInstanceKey(), value.getProcessDefinitionKey());
```

Comment:
❓ Why does this now have to be tenant-specific? The column family currently also exists as non-tenant specific. Note that the process instance and process definition keys are already fully unique, and their referenced values are tenant-specific.

Reply:
Because with this changes I added a fk to the process definition. The key of `processInstanceKeyByProcessDefinitionKeyColumnFamily` now contain also the tenant id. Is this make sense to you ?

Reply:
💭 Ah I see now. The `PROCESS_CACHE` column family contains the process definitions as a tenant aware key (tenant id is prefixed). So, in order for the `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY` column family to reference the `PROCESS_CACHE` we must also prefix the `processDefinitionKey` in the `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY` column family.

❌ This means we need to migrate the existing data in the `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY` column family and prefix it in a similar way as we did with the multi-tenancy migrations.

Reply:
I'd strongly suggest not migrating data this way, especially if it's not strictly required. We already had issues with the multi-tenancy migrations and adding another one is only going to make this worse.

Reply:
Yes but there should almost always be ways to avoid data migrations, for example by being backwards compatible with old schemas, by migrating data incrementally as needed or other strategies.
While we did make improvements to he way we run migrations to unblock us, these migrations inevitably require a lot of resources for large runtime use cases because our migrations are (at least) linear in the amount of data, here it would be the number of process instances.

> If problems are left, what can we do to overcome those problems in the future?
I'd suggest using a different strategy to implement this feature. For example (and if I understand this PR correctly), we could implement it such that new entries _are_ written with the tenant reference while old entries either do not benefit from the consistency checks or have to check all tenants for a matching process instance.

Reply:
>new entries are written with the tenant reference while old entries either do not benefit from the consistency checks or have to check all tenants for a matching process instance.

That sounds hard to maintain, and we explicitly chose not to take this route with multi-tenancy for that reason. The benefit of migrating the data is to avoid having to maintain two different ways of storing the data.

I guess the trade-off here is (pick 2): 
1. update performance, i.e. shorter or longer downtime of a partition during update
2. data consistency safety for process instance -> process definition
3. state code maintainability

Not sure what's best though. My gut says 2 and 3, as 1 is a one-time downside/upside, but 2 and 3 are both long-term.

Reply:
Since with 8.5 we run migration only on version upgrade, I think adding a new one has not a big impact for customer on daily basis. As already mentioned, the impact is on version upgrading. Might be worth to invest more time in making migration even more "smart". For this case, a migration should run only if updating from 8.5.x to 8.6.x

In general, I'm in favour of a longer downtime if we improve data consistency

Reply:
> update performance, i.e. shorter or longer downtime of a partition during update

Past experience was not that the update was "slow" but that it crashed brokers. At the very least I'd suggest you test new migrations on some large runtime state setups to ensure that they can finish within some memory bound.

Reply:
Am I right that we want to run another round of migrations for all existing PIs for introducing this FK which by default has no effect as consistency checks are disabled by default? These migrations will affect all users and not just those actually using consistency checks, right?

If so I would conclude the cost-benefit of this change is actually not worth doing the change at all and we rather not introduce the FK. The only compromise I see is implementing this check without a migration as consistency checks have a significant performance check anyway.

Reply:
>has no effect as consistency checks are disabled by default

Might make sense to reverse that at some point, making data safety the default, and allowing users to opt-out for performance

Reply:
> In general, I'm in favour of a longer downtime if we improve data consistency

By default users will get the downtime but not the latter, right? Also we embrace here a potential large migration transaction state risking to hit https://github.com/camunda/zeebe/issues/14975 for all users of which only a fraction has consistency checks and MT enabled.

I get we at some point have no other option but to migrate but in the context of this change I see more risks and downsides for the majority of users with a migration compared to implementing the check with a higher performance impact for a smaller subset of users that don't seem to have high performance demands anyway if checks are enabled.

Reply:
Ok, so as we don't yet have consistency checks enabled by default, I agree that we shouldn't add this foreign key because the downside (downtime during upgrade) doesn't outweigh the upside (data safety only to those that have consistency checks enabled, i.e. few).

Reply:
Should I close this PR then ?

Reply:
Ok with me. What about you, @megglos?